lib/backends/webgl/ops/conv.ts

```diff
@@ -1,16 +1,153 @@
+    const xShape = inputs[0].dims.slice();
+    const wShape = inputs[1].dims.slice();
+    const outputChannelsPerGroup = wShape[0] / this.group;
+    // if kernelShape is not specified in the attributes of this op, infer it from the weight tensor dims
+    if (this.kernelShape.length === 0) {
+      for (let i = 2; i < wShape.length; ++i) {
+        this.kernelShape.push(wShape[i]);
+      }
+    }
+    PoolConvUtil.adjustPadsBasedOnAutoPad(
+        inputs[0].dims, this.strides, this.dilations, this.kernelShape, this.pads, this.autoPad);
+    Logger.verbose(
+        'Conv',
+        `autpPad:${this.autoPad}, dilations:${this.dilations}, group:${this.group}, kernelShape:${
+            this.kernelShape}, pads:${this.pads}, strides:${this.strides}`);
+    const outputShape = WebGLConv.calcOutputShape(xShape, wShape, this.dilations, this.pads, this.strides);
+    const shaderSource = `
+    const ivec2 strides = ivec2(${this.strides[0]}, ${this.strides[1]});
+    const ivec2 pads = ivec2(${this.pads[0]}, ${this.pads[1]});
+
+    void main() {
+      ivec4 coords = getOutputCoords();
+      int batch = coords.x;
+      int output_channel = coords.y;
+      ivec2 xRCCorner = coords.zw * strides - pads;
+      int group_id = output_channel / ${outputChannelsPerGroup};
+
+      float dotProd = 0.0;
+      for (int wInChannel = 0; wInChannel < ${wShape[1]}; wInChannel++) {
+        int input_channel = group_id * ${wShape[1]} + wInChannel;
+        for (int wHeight = 0; wHeight < ${wShape[2]}; wHeight++) {
+          int xHeight = xRCCorner.x + wHeight * ${this.dilations[0]};
+
+          if (xHeight < 0 || xHeight >= ${xShape[2]}) {
+            continue;
+          }
+
+          for (int wWidth = 0; wWidth < ${wShape[3]}; wWidth++) {
+            int xWidth = xRCCorner.y + wWidth * ${this.dilations[1]};
+            if (xWidth < 0 || xWidth >= ${xShape[3]}) {
+              continue;
+            }
+
+            float xVal = getX(batch, input_channel, xWidth, xHeight);
+            float wVal = getW(output_channel, wInChannel, wWidth, wHeight);
+            dotProd += xVal*wVal;
+          }
+        }
+      }
+      ${processBias}
+      outputColor = vec4(dotProd, .0, .0, .0);
+    }
+`;
+    return {
+      inputLayouts: inputs.map(t => handler.getOrCreateTextureLayout(t)),
+      outputLayout: handler.createTextureLayoutFromShape(outputShape),
+      samplers: hasBias ? ['X', 'W', 'Bias'] : ['X', 'W'],
+      shaderSource,
+      hasMain: true,
+    };
+  }
+
+  createRunData(handler: WebGLInferenceHandler, programInfo: ProgramInfo, inputs: Tensor[]): RunData {
+    const inputTDs = inputs.map((t, i) => handler.getOrCreateTextureData(t, programInfo.inputLayouts[i]));
+    return {
+      inputTextureDatas: inputTDs,
+      outputTextureData: handler.createTextureDataFromLayout(programInfo.outputLayout, inputTDs[0].tensor.type),
+      uniformData: {}
+    };
+  }
+}
+
+export class WebGLUnpackedConv extends Conv {
```

Comment:
why `WebGLUnpackedConv` doesn't implement `WebGLOperator` as `WebGLUnpackedGroupedConv` do?

Reply:
yes, WebGLUnpackedConv consists of two kernels (im2col + matmul), which is not supported by WebGLOperator interface.

Reply:
ok. make sense.